python/src/main/java/org/apache/apex/malhar/python/base/BasePythonExecutionOperator.java

```diff
@@ -0,0 +1,591 @@
+   *  completely booted up in memory. The time taken to boot the interpreter depends on the libaries that are loaded etc
+   * @param context The Operator context
+   */
+  @Override
+  public void setup(Context.OperatorContext context)
+  {
+    super.setup(context);
+    apexPythonEngine = initApexPythonEngineImpl(context);
+    Map<PythonInterpreterConfig,Object> preInitConfigurations = getPreInitConfigurations();
+    apexPythonEngine.preInitInterpreter(preInitConfigurations);
+    apexPythonEngine.startInterpreter();
+    apexPythonEngine.postStartInterpreter();
+    initPartitioner();
+    processPostSetUpPythonInstructions(apexPythonEngine);
+  }
+
+  @Override
+  public void teardown()
+  {
+    super.teardown();
+    apexPythonEngine.stopInterpreter();
+  }
+
+  @Override
+  public void beginWindow(long windowId)
+  {
+    super.beginWindow(windowId);
+    requestCounterForThisWindow = 0;
+    responseCounterForThisWindow = 0;
+    numNullResponsesPerWindow = 0;
+    currentWindowId = windowId;
+  }
+
+  @Override
+  public void endWindow()
+  {
+    super.endWindow();
+    if (responseCounterForThisWindow < requestCounterForThisWindow) {
+      LOG.info("Detected stragglers and configured flag/state for blocking at end of window is " +
+          blockAtEndOfWindowForStragglers);
+      if (blockAtEndOfWindowForStragglers) {
+        while (responseCounterForThisWindow < requestCounterForThisWindow) {
+          LOG.debug("Trying to emit all stragglers before the next window can be processed");
+          emitStragglers();
+        }
+      }
+    }
+  }
+
+  @Override
+  public Collection<Partition<BasePythonExecutionOperator>> definePartitions(
+      Collection<Partition<BasePythonExecutionOperator>> partitions, PartitioningContext context)
+  {
+    initPartitioner();
+    return partitioner.definePartitions(partitions,context);
+  }
+
+  @Override
+  public void partitioned(Map<Integer, Partition<BasePythonExecutionOperator>> partitions)
+  {
+    initPartitioner();
+    partitioner.partitioned(partitions);
+  }
+
+  @Override
+  public void beforeCheckpoint(long windowId)
+  {
+    accumulatedCommandHistory.clear();
+    accumulatedCommandHistory.addAll(getApexPythonEngine().getCommandHistory());
+    numStarvedReturnsPerCheckpoint = getApexPythonEngine().getNumStarvedReturns();
+  }
+
+  @Override
+  public void checkpointed(long windowId)
+  {
+    getApexPythonEngine().setNumStarvedReturns(0L);
+    numberOfRequestsProcessedPerCheckpoint = 0;
+  }
+
+  @Override
+  public void committed(long windowId)
+  {
+  }
+
+  public void processPostSetUpPythonInstructions(ApexPythonEngine pythonEngine) throws ApexPythonInterpreterException
+  {
+  }
+
+  @Override
+  public void handleIdleTime()
+  {
+    emitStragglers();
+  }
+
+  @Override
+  public Response processStats(BatchedOperatorStats stats)
+  {
+    Response response = new Response();
+    if ( apexPythonEngine == null) {
+      response.repartitionRequired = true;
```

Comment:
Since the engine is initialized in setup, how do we ever get here?

Reply:
Thanks for pointing this out. Removed the stats listener interface completely.